src/arrow_reader.rs

```diff
@@ -806,45 +805,78 @@ impl NaiveStripeDecoder {
 }
 
 pub struct Cursor<R> {
-    pub(crate) reader: Reader<R>,
-    pub(crate) root_data_type: RootDataType,
+    pub(crate) reader: R,
+    pub(crate) file_metadata: Arc<FileMetadata>,
+    pub(crate) projected_data_type: RootDataType,
     pub(crate) stripe_offset: usize,
 }
 
-impl<R> Cursor<R> {
-    pub fn new<T: AsRef<str>>(r: Reader<R>, fields: &[T]) -> Result<Self> {
-        let projected_data_type = r.metadata().root_data_type().project(fields);
+impl<R: ChunkReader> Cursor<R> {
+    pub fn new<T: AsRef<str>>(mut reader: R, fields: &[T]) -> Result<Self> {
+        let file_metadata = Arc::new(read_metadata(&mut reader)?);
+        let projected_data_type = file_metadata.root_data_type().project(fields);
         Ok(Self {
-            reader: r,
-            root_data_type: projected_data_type,
+            reader,
+            file_metadata,
+            projected_data_type,
             stripe_offset: 0,
         })
     }
 
-    pub fn root(r: Reader<R>) -> Result<Self> {
-        let data_type = r.metadata().root_data_type().clone();
+    pub fn root(mut reader: R) -> Result<Self> {
+        let file_metadata = Arc::new(read_metadata(&mut reader)?);
+        let data_type = file_metadata.root_data_type().clone();
         Ok(Self {
-            reader: r,
-            root_data_type: data_type,
+            reader,
+            file_metadata,
+            projected_data_type: data_type,
             stripe_offset: 0,
         })
     }
 }
 
-impl<R: Read + Seek> Iterator for Cursor<R> {
+impl<R: AsyncChunkReader> Cursor<R> {
+    pub async fn new_async<T: AsRef<str>>(mut reader: R, fields: &[T]) -> Result<Self> {
+        let file_metadata = Arc::new(read_metadata_async(&mut reader).await?);
+        let projected_data_type = file_metadata.root_data_type().project(fields);
+        Ok(Self {
+            reader,
+            file_metadata,
+            projected_data_type,
+            stripe_offset: 0,
+        })
+    }
+
+    pub async fn root_async(mut reader: R) -> Result<Self> {
+        let file_metadata = Arc::new(read_metadata_async(&mut reader).await?);
+        let data_type = file_metadata.root_data_type().clone();
+        Ok(Self {
+            reader,
+            file_metadata,
+            projected_data_type: data_type,
+            stripe_offset: 0,
+        })
+    }
+}
```

Comment:
Could reduce duplication here

I'd prefer without macro if possible (just easier to work with Rust code instead of macros in my opinion)